### DISCUSSION TITLE
Permissions issues - sshfs

### DISCUSSION THREAD

**Discussion by nicolileo** [ltsp/ltsp#361] [Q&A]
Hello,

First great job on ltsp, glad this project goes on!

We are dealing with a permission issue, and i think it's sshfs relevant :

When a user is creating a file or a directory, permissions are 644 and we want 770 or 660 because we are using groups.
On our network we have a directory (/srv/shares) shared with NFS and autofs to mount this shares.

We tried to modify the script /usr/share/ltsp/client/login/pamltsp to remove parameter "default_permissions", nothing happens.

Have you got an idea?

Thanks in advance, Regards

**Comment by sebastiendu** [comment]
Not sure this will solve it as I do not use NFS for shared folders:
create /etc/profile.d/some_name.sh with this contents:
<pre>
umask 2
</pre>
Then run ltsp image once as usual.

**Reply by nicolileo** [reply]
Hi, 
What do you use to share folders?
But i'm pretty sure it's not NFS relevant. Same permissions if i create a directory inside or outside nfs shares.
Ok thanks for the tip i will try it next week.
Cheers

**Reply by sebastiendu** [reply]
> What do you use to share folders?

I have been trying to share /srv through SSH using this sequence of commands:

<pre>
ssh-keygen -q -N '' -f /etc/ltsp/id_rsa
( echo -n 'from="192.168.42.0/24",command="/usr/lib/openssh/sftp-server" ';cat /etc/ltsp/id_rsa.pub ) >> .ssh/authorized_keys
vim /etc/ltsp/ltsp.conf
[clients]
FSTAB_SRV="server:/srv /srv fuse.sshfs rw,nosuid,nodev,allow_other,IdentityFile=/etc/ltsp/id_rsa,StrictHostKeyChecking=no,UserKnownHostsFile=/dev/null"

ltsp initrd
</pre>

It works: the server's /srv folder is mounted on the clients. But I could not fix two problems:

1. All created files belong to root.
2. Created files are not writable by the group. The group is the right group by the write permission is not set. The umask does not seem to be used.

So I tried something else:

<pre>
rm /etc/ltsp/id_rsa
vim /etc/ltsp/ltsp.conf
[clients]
FSTAB_SRV="server:/srv /srv fuse.sshfs rw,nosuid,nodev,noauto,user,StrictHostKeyChecking=no,UserKnownHostsFile=/dev/null"
POST_INIT_SRV="chmod 777 /srv"

ltsp initrd
</pre>
(except LTSP resets /srv permissions later in the boot process, I am not sure where to move this chmod to)
Once the chmod is done (manually), any user can mount /srv as the user, and permissions are as expected.

A few remaining problems:

1. The system should mount /srv at login and unmount /srv at logout (XDG autorun service ? pamltsp hooks ?)
2. A second concurrent user will not be allowed to mount /srv on the same computer.
3. What resets /srv permissions after my POST_INIT_SRV ?

**Reply by sebastiendu** [reply]
Now (un)mounting /srv from pamltsp, at the same time and the same way the user's home directory is (un)mounted.

```diff
--- /usr/share/ltsp/client/login/pamltsp.orig	2020-06-09 16:48:06.000000000 +0200
+++ /usr/share/ltsp/client/login/pamltsp.new	2021-01-05 18:38:20.250747801 +0100
@@ -149,6 +149,7 @@
         # It needs a few seconds to kill all the processes.
         # Logins will fail for those seconds, and work properly afterwards.
         grep -qs "$pw_dir fuse.sshfs" /proc/mounts && fusermount -u "$pw_dir"
+        grep -qs "/srv fuse.sshfs" /proc/mounts && fusermount -u /srv
         # Create an empty home dir if it's not there; nope, no skel for SSHFS
         if [ ! -d "$pw_dir" ] && [ "$MKHOMEDIR" != "0" ]; then
             mkdir -p -m 0755 "$pw_dir"
@@ -165,9 +166,11 @@
         # fuse3 defaults to nonempty and doesn't accept it
         command -v fusermount3 >/dev/null ||
             sshfs_params="$sshfs_params,nonempty"
-        if "${_SELF%/*}/ssh-askpass" |
+        p=$("${_SELF%/*}/ssh-askpass")
+        if echo "$p" |
             sshfs -o "$sshfs_params" "$@" "$pw_name@$SERVER:" "$pw_dir"
         then
+            echo "$p" | sshfs -o "$sshfs_params" "$@" "$pw_name@$SERVER:/srv" "/srv"
             success=1
         fi
     fi
@@ -255,6 +258,7 @@
         # If no user processes are running, unmount it
         if [ "$(pgrep -cu "$pw_name")" = "0" ]; then
             fusermount -u "$pw_dir"
+            fusermount -u "/srv"
             return 0
         fi
     done
```

I just need to move it to another place because it would be overwritten when upgrading ltsp.

**Reply by alkisg** [reply]
Note that this will break when a user runs `su - otheruser`, i.e. when 2 simultaneous logins happen, as then two users will try to mount the same global directory (/srv).

**Comment by alkisg** [comment]
Regarding the 644 etc permissions, you need to set umask either from PAM or from .profile, for example:

```shell
$ umask 022; touch one; umask 002; touch two; ls -l one two
-rw-r--r-- 1 alkisg alkisg 0 Dec  31 15:59 one
-rw-rw-r-- 1 alkisg alkisg 0 Dec  31 15:59 two
```

Regarding the /srv mode, the permissions of the mount point are replaced by the ones of the remote directory. So when you mount server:/share over e.g. /srv/mountpoint, you get the permissions of the share.

Regarding shared directories over SSHFS, some quick thoughts.

Method 1would be to have a global share mounted from e.g. /etc/fstab. This usually means that the private SSH keys would exist in the LTSP image, which then means that it's unsafe as anyone can read them. Which is almost the same as just using a much easier, stabler and faster NFS share.

Method 2 would be to mount it per user, after login. This can be implemented like this:

1. Create a "shared" user on the server, with home=/home/shared (or the shared dir).
2. Run ssh-keygen for that user, and copy id_rsa.pub to authorized_keys so that he can login passwordlessly.
3. Copy the shared user's private id_rsa to every user's ~/.config/ltsp/id_rsa path, either via a `for` loop or via `/etc/skel` for new users.
4. Create a /usr/local/bin/mate-session script, that replaces the desktop environment session, runs after /home/username is mounted over SSHFS, and contains something like the following:

```shell
#!/bin/sh

mkdir -p shared
sshfs -o IdentityFile=~/.config/ltsp/id_rsa shared@server: shared
/usr/bin/mate-session "$@"
fusermount -u shared
```

Note that in method 2, it's best to mount it under ~/share and not in a general /srv/share directory, because LTSP clients support multiple logins, and it would be hard to manage the mounts/unmounts then.

Happy new year to all!

**Reply by nicolileo** [reply]
Hi,

sorry for the late answer and happy new year to all,

> Method 1would be to have a global share mounted from e.g. /etc/fstab. This usually means that the private SSH keys would exist in the LTSP image, which then means that it's unsafe as anyone can read them. Which is almost the same as just using a much easier, stabler and faster NFS share.

Alkisg, can you explain the method1 please? I think we can manage to mount the shared folder from /etc/fstab but is it the only thing to do (we have already change all users .profile to fit the umask we want)? Actually i don't understand the last part of your message as we are using a NFS share...

Thanks!

**Comment by alkisg** [comment]
Let me list all the methods as bullets and we can talk more about the one that you choose.

1. People that trust their LAN can use NFSv3 for everything; rootfs, /home and /srv/share, using FSTAB_x options in ltsp.conf. It's the easiest, most stable and fastest. It's not secure, so it's not the default for /home.
2. Others can use the default SSHFS for secure /home, and NFS for /srv/share, if the shared folder isn't security-sensitive, using an FSTAB_x option in ltsp.conf.
3. Some might put an SSHFS private key in the LTSP image and use SSHFS for /srv/share. This too can be done with an FSTAB_x option. But this isn't really secure because anyone in the LAN can read the private SSHFS key.
4. Finally, it's possible to use the login password or SSH keys (/home/user/.ssh/id_rsa) to do additional SSHFS mounts securely when logging in:
   1. Either in some path that contains $USER, so that it's different per user, e.g. /home/user/share,
   2. Or in some global path like /srv/share.

The 4.2 method is the hardest one to do properly; the 4.1 is a bit easier. Both can be implemented with pam_mount or with the session override option that I described above. Unfortunately I have no time now to properly document all the quirks that those might have.

**Reply by nicolileo** [reply]
Hello, 

We can't manage to do what we want on fresh install
reminder : we want 660 permissions on file and folder creation (umask 002)

On a ltsp client : 

$ umask 022; touch one; umask 002; touch two; ls -l one two
-rw-r--r-- 1 root nico  0 Jan  7 15:42 one
-rw-r--r-- 1 root nico  0 Jan  7 15:42 two

I use NFS to share a big folder but it's not NFS relevant as it's the same permissions if i create a directory inside or outside nfs shares.

Tried sebastiendu tip :

>  Not sure this will solve it as I do not use NFS for shared folders:
> create /etc/profile.d/some_name.sh with this contents:
> umask 2
> Then run ltsp image once as usual.

and of course modified .profile for each account -> NOK

Can we add something in ltsp.conf or try something like this? https://github.com/ltsp/ltsp/pull/368

Did we missed something obvious?

**Comment by nicolileo** [comment]
ok thanks for your answer, i will look into it

**Comment by sebastiendu** [comment]
Just submitted a pull request to add SSH-mounted shared directories : #368

**Comment by sebastiendu** [comment]
> $ umask 022; touch one; umask 002; touch two; ls -l one two
> -rw-r--r-- 1 root nico  0 Jan  7 15:42 one
> -rw-r--r-- 1 root nico  0 Jan  7 15:42 two
> 
> I use NFS to share a big folder but it's not NFS relevant as it's the same permissions if i create a directory inside or outside nfs shares.

Please show

* the server-side /etc/exports line
* the client-side /etc/fstab line

**Comment by alkisg** [comment]
@nicolileo, github supports "threaded discussions", but it's best NOT to use them, as they're confusing. In most cases it's best to reply at the bottom of the issue, not inline.

If the `umask` test commands aren't working properly for you, then you have a serious problem that other LTSP users don't have. Like @sebastiendu says, paste the relevant server exports and client fstab files, but also test the `umask` commands in a subdirectory of /tmp in the client, where I assume that the permissions shouldn't be affected by fstab.

**Comment by nicolileo** [comment]
You are so quick to respond guys!
As i'm not in front right now, i can only paste the lines you asked this afternoon.
Thanks!

**Comment by nicolileo** [comment]
Server side
$ cat /etc/exports.d/local.exports 

/srv/share/Commun 192.168.100.0/24(rw,sync,no_subtree_check)

Client side : 
$ cat /etc/fstab

192.168.100.4:/share /srv/share/Commun nfs auto,rw, 0  0

Like you said Alkisg, umask command works fine in /tmp but umask doesn't work fine in /home/USER and isn't affected by fstab too.

**Reply by sebastiendu** [reply]
> 192.168.100.4:/share /srv/share/Commun nfs auto,rw, 0 0
> 
> Like you said Alkisg, umask command works fine in /tmp but umask doesn't work fine in /home/USER and isn't affected by fstab too.

When you test umask, are you under /home/USER or /srv/share/Commun ?

**Comment by nicolileo** [comment]
under /home/USER

**Comment by alkisg** [comment]
Run the following commands as a user on an LTSP client, and send us their output:

```shell
grep home /proc/mounts
cd ~/.cache
umask
umask 022; touch one; umask 002; touch two; ls -l one two
touch three; chmod 777 three; ls -l three
```

**Comment by chenaille** [comment]
Sorry when i mentionned home/USER, i meant /srv/home/USER

```shell
lileo@ltsp113:~$ grep home /proc/mounts
lileo@server: /srv/home/lileo fuse.sshfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
lileo@ltsp113:~$ cd .cache/
lileo@ltsp113:~/.cache$ umask
0022
lileo@ltsp113:~/.cache$ umask 022; touch one; umask 002; touch two; ls -l one two
-rw-r--r-- 1 lileo lileo 0 janv.  8 16:30 one
-rw-r--r-- 1 lileo lileo 0 janv.  8 16:30 two
lileo@ltsp113:~/.cache$ touch three; chmod 777 three; ls -l three
-rwxrwxrwx 1 lileo lileo 0 janv.  8 16:31 three
```

**Comment by chenaille** [comment]
sorry for the mess...

**Comment by alkisg** [comment]
And the output of these commands?

```shell
which umask
getent passwd $USER
```

Why have you placed /home under /srv? That way, on the client, it will be mounted in a different directory, and some applications will have problems locating the proper paths...

**Comment by chenaille** [comment]
```shell
lileo@ltsp112:~$ which umask
lileo@ltsp112:~$ getent passwd $USER
lileo:x:1000:1000:lileo,,,,umask=007:/srv/home/lileo:/bin/bash
```
Yes perhaps it's a mistake but the server is made this way : 
/dev/md0    28G   7,9G   19G   31%  /
/dev/md1    1,8T   410G  1,3T  25%  /srv

**Comment by alkisg** [comment]
And if you login as lileo on the server, and you run the same test umask commands under /srv/home/lileo/.cache, you see different results? You're using `bash` everywhere, right?

Why does the user entry in passwd have "umask=007" in the gecos field? It even has more commas than it should. Try to remove that, run ltsp initrd, and see if it fixes things.

I would mount /dev/md1 under /home, give all users a proper /home/username (and update passwd accordingly), create a /home/srv directory, and symlink /srv to /home/srv.

Anyway it's a mystery why `umask` doesn't behave properly for you, it does for all other LTSP users. If you want, you may come in IRC for live troubleshooting.

**Comment by nicolileo** [comment]
Hi, 

> And if you login as lileo on the server, and you run the same test umask commands under /srv/home/lileo/.cache, you see different results? 

yes, normal results

> You're using bash everywhere, right?

yes

> Why does the user entry in passwd have "umask=007" in the gecos field?

I can't see the commas??

> Anyway it's a mystery why umask doesn't behave properly for you

Agree, and the weirdest is that we have the same behavior on a fresh install of debian 10...

Any suggestion for a bypass?

Thanks anyway

**Comment by alkisg** [comment]
> I can't see the commas??

Compare yours with mine:

```shell
lileo:x:1000:1000:lileo,,,,umask=007:/srv/home/lileo:/bin/bash
administrator:x:1000:1000:Administrator,,,:/home/administrator:/bin/bash
```

> Agree, and the weirdest is that we have the same behavior on a fresh install of debian 10...

Try SSHFS without LTSP. Go to the server, and run:

```shell
mkdir /tmp/sshfs
sshfs lileo@localhost: /tmp/sshfs
touch /tmp/sshfs/one etc etc
```

> Any suggestion for a bypass?

IRC!

**Comment by nicolileo** [comment]
ok see u on wednesday
